[gsoc] add feature to display transitive dependencies

nakul9597 opened the pull request:
This PR adds a feature that allows users to look at all the transitive dependencies of a gem.
adding a "/dependencies" after the gem name or version number (for specific version) will render a view that shows all the transitive dependencies. 

Example: two ways of calling the view
http://localhost:3000/gems/rails/dependencies
http://localhost:3000/gems/rails/versions/0.8.0/dependencies

dwradcliffe commented:
I haven't looked at the code yet, but "recursive" is a red flag for performance. What is the response time for large trees of dependencies?

nakul9597 commented:
The time for the whole list took around 6 seconds when it was not cached and 1 second when it was cached. We've planned to change the design by not displaying all the transitive dependencies but only the first 5 or so dependencies and when the user unfolds the last one we will make a request to display more dependencies.

nakul9597 commented:
I've updated the pr to display dependencies preceded by an arrow that when clicked loads up its transitive dependencies. As we keep loading the transitive dependencies two things happen:
  1. If a dependency is already displayed and it comes again it can't be expanded and has a red colour while the dependencies that are opened changes to an orange colour.
  2. On the right side of the page, the list of dependencies keep getting updated.

I've also given an option to expand and retract the transitive dependencies.
This new code does two ajax requests: 
  1. expanding and retracting the transitive dependencies.
  2. updating list of transitive dependencies.

To be done: Add test cases for the code i have written and change naming conventions if necessary.

I've attached an image of how the view looks. Please tell me for any changes in design also.
https://user-images.githubusercontent.com/16337625/60026770-d77e3700-96b9-11e9-8f39-4f6a1cc39468.png

sonalkr132 commented on app/views/transitive_dependencies/_dependencies.html.erb:
```diff
@@ -0,0 +1,22 @@
+          <div><span class="deps_item deps_item--duplicate"><%= rubygem.name %> <%= dependency.clean_requirements %></span></div>
+        <% else %>
+          <ul class="deps"><li><span id="<%=rubygem.name%>-toggle"><%= link_to rubygem_find_by_name_path(gem_name: rubygem.name,format: :js), remote: true do %>
+          <span class="caret"></span><% end %></span><span id="<%=rubygem.name%>-title"><strong><%= rubygem.name %></strong> <%= dependency.clean_requirements %></span>
```
Please remove strong from here. Change in font size looks jumpy.

nakul9597 replied:
Okay. but I thought the gem name should stand out from the version number. Moreover I used the same way it was used to show dependency in gem web page.
https://github.com/rubygems/rubygems.org/blob/master/app/views/rubygems/_dependencies.html.erb#L9

This is without strong
https://user-images.githubusercontent.com/16337625/60722560-f5b42480-9f4e-11e9-8974-139d440a298e.png

This is with strong
https://user-images.githubusercontent.com/16337625/60722616-18ded400-9f4f-11e9-9d1b-c1a38c8913b3.png

Please let me know if I should remove it.

sonalkr132 replied:
okay. it's not about the use of strong tag then. On clicking the toggle arrow, the font size of the gem name increases. please remove that.

sonalkr132 commented on app/controllers/versions_controller.rb:
```diff
@@ -11,4 +12,49 @@ def show
   end
+
+  def dependencies_show
+    params[:version_id] ? latest_version_by_slug : latest_version
```
When would `version_id` be absent?

nakul9597 replied:
For this webpage we need to find latest_Version
http://localhost:3000/gems/rails

For this we need to find_by_slug since it can have different dependencies from the latest version.
http://localhost:3000/gems/rails/versions/5.2.2

nakul9597 commented on app/controllers/versions_controller.rb:
```diff
@@ -11,4 +12,49 @@ def show
+  end
+
+  def transitive_dependencies_show
+    @latest_version = find_dependencies_by_gem_name(params[:gem_name])
```
Okay. Thanks, I have done that.

nakul9597 commented on app/views/rubygems/show.html.erb:
```diff
@@ -41,6 +41,11 @@
     <%= render :partial => "rubygems/dependencies", :locals => { :dependencies => @latest_version.dependencies.development } %>
+    <div class = "l-half--r">
+    <div class="gem__dependency__list">
+          <%= link_to "dependency list →", rubygem_version_path(@rubygem, @latest_version.number)+"/dependencies", class: "dependency__list__button" %>
```
I have changed this. Thanks.

nakul9597 commented on app/views/transitive_dependencies/_dependencies.html.erb:
```diff
@@ -0,0 +1,22 @@
+        <% if @gem_deps_names.include?(rubygem.name) %>
+          <div><span class="deps_item deps_item--duplicate"><%= rubygem.name %> <%= dependency.clean_requirements %></span></div>
+        <% else %>
+          <ul class="deps"><li><span id="<%=rubygem.name%>-toggle"><%= link_to rubygem_find_by_name_path(gem_name: rubygem.name,format: :js), remote: true do %>
```
I just realised when you pointed it out. Thanks for that. I had to change the padding a little bit but now it looks the same.

nakul9597 commented on app/views/transitive_dependencies/_dependencies.html.erb:
```diff
@@ -0,0 +1,22 @@
+    <% dependencies.each do |dependency| %>
+      <% if rubygem = dependency.rubygem %>
+        <% if @gem_deps_names.include?(rubygem.name) %>
+          <div><span class="deps_item deps_item--duplicate"><%= rubygem.name %> <%= dependency.clean_requirements %></span></div>
```
we need "transitive_dependencies/dependencies" because I am rendering it with different dependencies during the ajax requests(loading new dependencies)
https://github.com/rubygems/rubygems.org/pull/2029/commits/3163ce1c2ce021faba96cb44f41761f4fb27bd1b#diff-dd903fc5354c928ae886bbf9706832deR4

I actually added the link to the gem in the list of dependencies on the right hand side. I can add it here too.

sonalkr132 replied:
dependencies list on the left side feels redundant. Is the `transitive_dependencies/dependencies` providing any additional info to the user which doesn't already exist on the page?

nakul9597 replied:
Do you mean the list on the right side(dependency list)? https://user-images.githubusercontent.com/16337625/60026770-d77e3700-96b9-11e9-8f39-4f6a1cc39468.png

Is it redundant? I can remove it if so. I just added it for users to get a list in alphabetical order.

nakul9597 commented on app/controllers/versions_controller.rb:
```diff
@@ -11,4 +12,49 @@ def show
+    end
+  end
+
+  def arrow_toggler
```
Yes. I've changed that. Thank you

sonalkr132 commented on config/routes.rb:
```diff
@@ -152,7 +152,11 @@
+      resources :versions, only: %i[show index] do
+        get '/dependencies', to: 'versions#dependencies_show'
+      end
+      get "/find-by-name", to:  "versions#transitive_dependencies_show", constraints: { format: :js }
```
please use a better name for this route. `find-by-name` sounds more like a function name than a route.

nakul9597 commented:
updates 
- made finding existing dependencies on page client side
- made toggling client side
- removed _dependencies.html.erb and extra ajax requests
- cleaner design

https://user-images.githubusercontent.com/16337625/61372263-4bc08a80-a8b4-11e9-8a9d-0f4c4cf94e51.png

sonalkr132 commented on app/controllers/versions_controller.rb:
```diff
@@ -11,4 +12,25 @@ def show
+    @gem_details = @latest_version.dependencies
+    @deps_names = @gem_details.collect { |dependency| dependency.rubygem.name }
+    respond_to do |format|
+      format.js { render "transitive_dependencies/show" }
```
Hey, lets change this to return JSON response, move the ` show.js.erb` code to `assets/javascript`. It is better to stick with the asset pipeline than compiling JS on request.

sonalkr132 commented on app/views/transitive_dependencies/show.js.erb:
```diff
@@ -0,0 +1,32 @@
+var gem_deps = <%=@deps_names.inspect.html_safe%>;
+
+$.each(gem_deps, function (idx,dep) {
+	if(all_deps.includes(dep)==true) {
```
Showing duplicate dependency at the bottom looked nicer than this:

![Screenshot from 2019-07-19 10-42-17](https://user-images.githubusercontent.com/7680662/61510782-fe5a2f80-aa11-11e9-86fd-a4bd36c047c6.png)

simi replied:
Is `==true` part really needed?

sonalkr132 commented on app/assets/stylesheets/modules/gem.css:
```diff
@@ -236,3 +236,25 @@
+}
+.dependency__list__button.disabled:hover {
+  color: #9da2ab
+}
```
Don't forget the new line.

sonalkr132 commented:
We will be adding selenium webdriver for system tests of JS features. Please read about how would be go about configuring it: https://github.com/teamcapybara/capybara#selenium

nakul9597 commented on app/views/transitive_dependencies/show.html.erb:
```diff
@@ -0,0 +1,24 @@
+        <ul class="deps">
+          <li><span id="<%=rubygem.name%>-toggle"><%= link_to rubygem_dependency_details_path(rubygem_id: rubygem.name,format: :js), remote: true do %>
+          <span class="caret"></span><% end %></span>
+          <%= link_to rubygem_path(id: rubygem.name), target: :_blank do %><span data-id="<%=rubygem.name%>" id="<%=rubygem.name%>-title" class="deps_item"><%= rubygem.name %> </span><%end%>
```
For that I'll have to do queries for each dependencies separately.
- temp_rubygem = Rubygem.find_by_name(gem_name) //find rubygem by name(dependency name)
- temp_rubygem.versions.most_recent["number"] //find versions.most_recent

two queries for each dependency. Is that okay?

sonalkr132 replied:
I think eager loading versions and dependencies table would be more efficient. Something like [this](https://github.com/rubygems/rubygems.org/blob/a7d0b57fff414223f468de79b6785f38d8b01e3d/app/models/concerns/rubygem_searchable.rb#L84).

simi commented on app/views/transitive_dependencies/show.js.erb:
```diff
@@ -0,0 +1,32 @@
+var all_deps = $('.deps_item').map(function() {
+	return $(this).attr('data-id');
```
I think indent by 2 spaces will fit in this file as well since it is used in all another templates.

nakul9597 replied:
Yes. I've changed it. Thank you for that.

simi commented on config/routes.rb:
```diff
@@ -153,7 +153,10 @@
         defaults: { format: :js }
-      resources :versions, only: %i[show index]
+      resources :versions, only: %i[show index] do
+        get '/dependencies', to: 'versions#dependencies_show'
```
I think we can keep controllers small and keep them "RESTful" -> create new controller for dependencies and also for transitive dependencies having only show action.

nakul9597 replied:
Hi, Thanks. I've changed it.

nakul9597 commented:
updates
- moved show.js.erb to asset pipeline and transitive_dependencies_show to return json to ajax call
- design now shows duplicate dependencies in the bottom
- add controllers for dependencies and transitive_dependencies show, removed from versions_controller.rb
- add module to find the best version of gem based on requirement
- add basic tests for successful controller requests.

nakul9597 commented:
Updates
- removed self written module and used rubygem files to reuse code
- added test case for js

sonalkr132 commented on app/controllers/dependencies_controller.rb:
```diff
@@ -0,0 +1,18 @@
+      dep_resolver(
+        dependency.rubygem.name,
+        dependency["requirements"],
+        Version.numbers_from_gemid(dependency["rubygem_id"]).to_a.uniq
```
Perhaps this method belongs in `Rubygems` (has multiple version number), where you could use `public_versions`. 
`uniq` or `to_a` seems unnecessary, why do we need them?

nakul9597 replied:
I actually don't need that code. I can delete it. I used that because while fetching some had multiple versions with same name(different platform ruby and java) and number.

nakul9597 replied:
Should I only fetch where platform is ruby?

nakul9597 replied:
Question: If two versions have same number where one is ruby and other is java. If both satisfy the condition of requirement, which one should I choose?

sonalkr132 commented on app/controllers/concerns/requirements_version.rb:
```diff
@@ -0,0 +1,15 @@
+require "rubygems/dependency"
```
Much clener now :crystal_ball: Thanks @nakul9597

nakul9597 replied:
Thank you for that help :D

sonalkr132 commented on app/controllers/concerns/latest_version.rb:
```diff
@@ -5,5 +5,9 @@ module LatestVersion
     end
+
+    def latest_version_by_slug
+      @latest_version = Version.find_by!(full_name: "#{params[:rubygem_id]}-#{params[:version_id]}")
```
Can you please also look into including platform here? [nokogiri java](https://rubygems.org/gems/nokogiri/versions/1.10.3-java) has concourse and simplecov as dependencies but [nokogiri ruby](https://rubygems.org/gems/nokogiri/versions/1.10.3) doesn't.

sonalkr132 commented on app/views/dependencies/show.html.erb:
```diff
@@ -0,0 +1,31 @@
+<% @subtitle = @latest_version.try(:slug) %>
+<% if @dependencies.present? && @latest_version.indexed %>
+  <div class="l-full--l">
+    <% @dependencies.each_with_index do |dependency,idx| %>
```
It would be nicer if we first show runtime and then development dependencies, also include this type info in the list.
https://guides.rubygems.org/patterns/#declaring-dependencies

nakul9597 commented:
Updates
- split view of dependencies based on whether it is runtime or development
- display requirements along with name and version
- added chrome stable to run js test

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,85 @@
+    }
+    else {
+      //ver_num = versions[idx]
+      var link = "<a href=\"javascript:ajaxDepOnClick('"+dep+"','"+ver_num+"')\"><span class='caret'></span></a>";
```
Please use `click()` instead of inline js. add `dep` and `ver_num` as `data-attr` and get their values in the click function instead of passing them as arguments.

sonalkr132 commented on app/controllers/transitive_dependencies_controller.rb:
```diff
@@ -0,0 +1,33 @@
+
+  def json_return
+    {
+      rubygem_id: params[:rubygem_id],
```
We don't need to send `rubygem_id`, read it from a data-attr.

sonalkr132 commented on app/views/dependencies/_dependencies.html.erb:
```diff
@@ -0,0 +1,31 @@
+<% if @dependencies.key?(scope) %>
+  <% if scope == "development" %>
+    <div id="<%=params[:rubygem_id]+scope%>"><span class="scope"><%= scope.titlecase %> :</span>
```
Using `params` (user input) directly is very dangerous, please use `@latest_version.rubygem.name`.

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,85 @@
+  }).get();
+
+  if (resp.run_deps.length != 0){
+    $("#"+gem_id+'runtime').find(".deps_scope").before("<span class='scope scope--expanded'>Runtime :</span>");
```
Why are we reinserting here?

nakul9597 replied:
That is where i'm inserting the name of scope (development or runtime)

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,85 @@
+  gem_id = gem_id.replace(/\./, "\\.")
+
+  var all_deps = $('.deps_item').map(function() {
+    return $(this).attr('data-id');
```
Use a more descriptive name than `data-id`.

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,85 @@
+    dep = dep_details[0]
+    ver_num = dep_details[1]
+    req = dep_details[2]
+    if(all_deps.includes(dep)) {
```
In the following case, rake 11.3.0 was not toggleable as we have shown rake 12.3.2, which isn't right. 12.3.2 could have different dependencies than 11.3.0
![Screenshot from 2019-08-16 00-08-12](https://user-images.githubusercontent.com/7680662/63117765-13e35a80-bfba-11e9-83e6-43f2418b733f.png)

This condition of already shown is making everything complicated. I think we should remove this check and always allow further toggle.

nakul9597 replied:
Okay. I'll change that or I can check name and version. Not a lot of change in code.

sonalkr132 commented on test/test_helper.rb:
```diff
@@ -51,6 +52,13 @@ class ActionDispatch::IntegrationTest
+  include Capybara::Minitest::Assertions
+
+  # Reset sessions and driver between tests
+  teardown do
```
This is needed only in `TransitiveDependenciesTest`, please move it there.

nakul9597 commented:
Updates
- used click() instead of inline js
- changed naming of some data-attributes
- changed use of params variables
- moved use of teardown in test_helper to test file

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,67 @@
+    var deps_list = deps_run+deps_dev;
+    new_gems.append("<ul class='deps'><li>"+toggle_link+link_to_gem+deps_list+"</li></ul>");
+  });
+  if (scope == "development"){
```
Why do we need this?

nakul9597 replied:
To toggle development's arrow to show side ways (not down like runtime) and hide all the development dependencies.

nakul9597 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,67 @@
+  });
+})
+
+function ajaxDepActions(resp,gem_id,current) {
```
No, I tried. Unless I'm in that same function $(this) is undefined.

nakul9597 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,67 @@
+function ajaxDepActions(resp,gem_id,current) {
+
+  // when we have gems like http_parser.rb, we need to replace . with \. to access any ids with the name
+  gem_id = gem_id.replace(/\./, "\\.")
```
I actually don't need it because I removed all the ids so i'm deleting it

nakul9597 commented:
Updates

- update self to connect_src in csp.rb
- correct naming   
- refactor js file further
- add error case for ajax request
- remove trans_dep_ctr.rb and made json to return from dep_ctr.rb

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,90 @@
+  $(this).next().next().toggleClass("deps_toggle")
+})
+
+function errorHandler(jqXHR, exception) {
```
We don't need all these cases or alert. Any error would update the relevant toggle div with a generic message like `request failed. please reload the page and try again`.

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,90 @@
+
+  scope_display(current, gem_id, resp.run_deps, "runtime")
+  scope_display(current, gem_id, resp.dev_deps, "development")
+
```
Remove the new line here and line 18.

sonalkr132 commented on app/assets/stylesheets/modules/dependencies.css:
```diff
@@ -0,0 +1,109 @@
+    border-right: 0.2em solid grey;
+    border-top: 0.2em solid grey;
+    transform: rotate(45deg);
+    margin-right: 0.5em;
```
Remove comments and fix the indentation.

sonalkr132 commented on app/views/dependencies/_dependencies.html.erb:
```diff
@@ -0,0 +1,27 @@
+        <%= link_to rubygem_version_path(rubygem_id: name, id: version), target: :_blank do %>
+          <span class="deps_item"><%= name %> <%= version %>
+          <span class='deps_item--details'> <%=req%></span></span>
+        <%end%>
```
space around `end`

sonalkr132 commented on app/views/dependencies/show.html.erb:
```diff
@@ -0,0 +1,11 @@
+
+    <%= render partial: "dependencies/dependencies", locals: { scope: "runtime" } %>
+    <%= render partial: "dependencies/dependencies", locals: { scope: "development" } %>
+
```
remove the new line here and line 6

nakul9597 commented:
Updates:
- remove redundant spaces
- change ajax error function to display generic error message

sonalkr132 commented on app/views/rubygems/show.html.erb:
```diff
@@ -41,6 +41,13 @@
     <%= render :partial => "rubygems/dependencies", :locals => { :dependencies => @latest_version.dependencies.runtime } %>
     <%= render :partial => "rubygems/dependencies", :locals => { :dependencies => @latest_version.dependencies.development } %>
+    <% if @latest_version.dependencies.present? && @latest_version.indexed %>
+      <div class = "l-half--r">
```
I think we need `l-half--l` here.

sonalkr132 replied:
Okay, `--l` seems to be left-aligned, let's remove this div. Nest everything from line 42 to single `<div class = "l-half--l">`, change top div in `_dependencies.html.erb` to `t-list__items` and add `push--s` to `link_to` classes.

sonalkr132 commented on app/views/dependencies/_dependencies.html.erb:
```diff
@@ -0,0 +1,27 @@
+    <div id="<%=@latest_version.rubygem.name+scope%>"><span class="scope"><%= scope.titlecase %> :</span>
+    <div class="deps_scope deps_toggle">
+  <% else %>
+    <div id="<%=@latest_version.rubygem.name+scope%>"><span class="scope scope--expanded"><%= scope.titlecase %> :</span>
```
Remove `if else` on scope and change span to following:
`<span class="scope <%= scope--expaned if scope == "development" %>">`.
Please also make sure there is space around there erb tag (<%= <> %>) everywhere.

sonalkr132 commented on app/views/dependencies/_dependencies.html.erb:
```diff
@@ -0,0 +1,27 @@
+  <% @dependencies[scope].each do |name,version,req| %>
+    <ul class="deps">
+      <li>
+        <span>
```
This span seems redundant.

nakul9597 replied:
Actually I need this span because I'll be changing the contents of this span after it is clicked once. I'll be changing the content to just toggle the div's to follow. (So that it doesn't cause ajax requests anymore.

nakul9597 replied:
I will be changing the contents of that span from .caret class (which I have now changed to deps_expanded-link) to the following for toggling.

![Screenshot from 2019-08-25 11-08-14](https://user-images.githubusercontent.com/16337625/63646036-b02eff00-c728-11e9-96f9-860862098aff.png)

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,71 @@
+function arrow_toggler(current) {
+  var toggler = "<span class='deps_expanded deps_expanded-down'></span>";
+
+  current.parent().click(function() {
```
`$(document).on('click', '.scope'` on line 77 registers the function for all elements, you don't need to do this again here.

nakul9597 replied:
This is for gem_name toggling. Used to toggle hiding both of the scopes and its dependencies. 

.scope toggling is to toggle showing and hiding only the dependencies of that particular scope.

nakul9597 replied:
.scope will toggle its class scope--expanded and its sibling deps_scope.
![Screenshot from 2019-08-25 11-17-59](https://user-images.githubusercontent.com/16337625/63646096-3566e380-c72a-11e9-9cc1-51c260e4d041.png)

nakul9597 replied:
While the arrow toggle should toggle it's parents class and its parents sibling's div classes.
The two div classes are used for each scope.
![Screenshot from 2019-08-25 11-21-30](https://user-images.githubusercontent.com/16337625/63646117-b7efa300-c72a-11e9-8f07-a7ee3defa49e.png)

![Screenshot from 2019-08-25 11-23-56](https://user-images.githubusercontent.com/16337625/63646122-deadd980-c72a-11e9-9a43-ef58886ce574.png)

sonalkr132 commented on app/assets/stylesheets/modules/dependencies.css:
```diff
@@ -0,0 +1,114 @@
+  user-select: none;
+}
+
+.caret::before {
```
`caret` and `caret::before` is same as `deps_expanded` and `deps_expanded::before` except color. You should be able to override color in `deps_expanded-down::before`. please remove this.

nakul9597 replied:
Thanks I've changed it. I renamed it as deps_expanded-link because I need to use it to click for ajax requests.

sonalkr132 commented on app/assets/stylesheets/modules/dependencies.css:
```diff
@@ -0,0 +1,114 @@
+  transition: 0.3s;
+}
+
+.gem_name{
```
Doesn't seem to be used.

sonalkr132 commented on app/assets/stylesheets/modules/gem.css:
```diff
@@ -236,3 +238,25 @@
+}
+.dependency__list__button.disabled:hover {
+  color: #9da2ab
+}
```
None of the changes in this file seem to be serving any purpose.

sonalkr132 commented on app/assets/stylesheets/modules/dependencies.css:
```diff
@@ -0,0 +1,114 @@
+.deps{
```
Space before `{` in all places.

sonalkr132 commented on app/views/dependencies/show.html.erb:
```diff
@@ -0,0 +1,9 @@
+<% @title = params[:rubygem_id] %>
+<% @subtitle = @latest_version.try(:slug) %>
+<span id="deps_all"></span>
```
What is this?

nakul9597 replied:
Oh no. I had used that for debugging. I'll remove it.

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,71 @@
+function scope_display(current, gem_id, deps,scope) {
+  if (deps.length != 0){
+    var new_gems = current.parent().next().next()
+    if (scope == "development") { new_gems = new_gems.next() }
```
Remove top div in following:
```
<div><div class="deps_scope"></div></div>
```
Use `deps_scope` selector with `closest` or `siblings` function instead of `current.parent().next().next()`. You can do away with `scope == "development"` check (and remove this function all together) if you use different selector for runtime and development scope/div.

nakul9597 replied:
The reason I have a div before that is because the <span class='scope scope--expanded'>(Runtime/development) :</span> comes before the deps_scope div and inside the top div.

![Screenshot from 2019-08-25 11-00-54](https://user-images.githubusercontent.com/16337625/63645981-ace74380-c727-11e9-9363-d8305d7cd73c.png)

nakul9597 replied:
I need that div for toggling scope and also the arrow

nakul9597 commented:
Updated with changes in class name (also click functions) and toggling.

sonalkr132 commented on .travis.yml:
```diff
@@ -18,6 +18,7 @@ addons:
       - elasticsearch
       - openjdk-8-jre-headless
   postgresql: "9.6"
+  chrome: stable
```
I think we need to update our `running tests` section with instruction for set up required for running selenium tests. Can you please send a separate PR for that?

nakul9597 replied:
I need to send a separate pr with the code that is required to run js tests?

sonalkr132 replied:
yes, please. not exactly code but instructions for packages to install. It would go [here](https://github.com/rubygems/rubygems.org/blob/master/CONTRIBUTING.md#environment-linux---debianubuntu).

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,70 @@
+  runtime_div.toggleClass('deps_toggle');
+  runtime_div.next().toggleClass('deps_toggle');
+  $(this).toggleClass('deps_expanded-down');
+})
```
@simi @eliotsykes 
Do you guys have opinions on this js we are adding? It feels a bit difficult to understand and `$(this).parent().next().next()` really erks me. I suggested using `closest` or `siblings` function instead of this, however that won't be phenomenally better either. We are adding integration tests so this should be fine?

@nakul9597 please add a semicolon at end of every line.

nakul9597 replied:
I've added a semicolon after each line. Sorry for that. Also should I change $(this).parent().next().next() ?

sonalkr132 commented on app/assets/javascripts/transitive_dependencies.js:
```diff
@@ -0,0 +1,70 @@
+  $.each(deps_names, function (idx,dep_details) {
+    dep = dep_details[0]
+    version = dep_details[1]
+    req = dep_details[2]
```
add var.

nakul9597 replied:
Thanks. I've done that.
